### PR TITLE
V2.0/make user identity optional

### PR DESCRIPTION
**Pull request by DyrellC**
# Description of change
Allow users to be created without a base `Identity` (as in a user reading a public channel or one using PSK's for strictly reading purposes).

#249

## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested
- `UserBuilder` test for adding `PSK`'s now conducted without the `Identity` input
- Examples remove `Identity` from `SubscriberC`

### CODE REVIEW

File: streams/src/api/user.rs [hunk 1]
```diff
@@ -942,7 +963,10 @@ impl<T> Debug for User<T> {
         write!(
             f,
             "\n* identifier: <{}>\n{:?}\n* PSKs: \n{}\n* messages:\n{}\n",
-            self.identifier(),
+            match self.identifier() {
+                Ok(id) => id.to_string(),
+                Err(_) => "User has no Identifier".to_string(),
```
Comment (arnauorriols) [inline streams/src/api/user.rs:968]: Maybe a nicer output would be to just not print `identifier: <>` if the user does not have one

File: streams/examples/full-example/scenarios/basic.rs [hunk 2]
```diff
@@ -260,7 +257,7 @@ pub(crate) async fn example<T: GenericTransport>(transport: T, author_seed: &str
             author
                 .subscribers()
                 .map(|s| {
-                    if s == subscriber_a.identifier() {
+                    if s == subscriber_a.identifier().unwrap() {
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply streams/examples/full-example/scenarios/basic.rs:260]: No particular reason, i'll change that after I merge the topics changes into the branch

File: streams/src/api/user.rs [hunk 3]
```diff
@@ -110,14 +117,25 @@ impl<T> User<T> {
     }
 
     /// User's identifier
-    pub fn identifier(&self) -> Identifier {
-        self.state.user_id.to_identifier()
+    pub fn identifier(&self) -> Result<Identifier> {
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply streams/src/api/user.rs:120]: Yeah it's the ergonomics of utility part that put me on the fence for this as well. Having to use `ok_or()?` everywhere really hurts the cleanliness of the `User` implementation. It's such a regularly used function that to keep things tidy `Result` made the most sense to me, but it is llike you said a little misleading in its intention vs its interpretation on reading. 

I think a happy medium point here would be: 
`User::Identity()` remains a `Result`, and i'll fetch the `Identifier` from that with the `to_identifier()` calls, and `User::Identifier()` will become `Option`

File: streams/src/api/user.rs [hunk 4]
```diff
@@ -1219,7 +1213,7 @@ impl<T> Debug for User<T> {
         write!(
             f,
             "\n* identifier: <{}>\n* topic: {}\n{:?}\n* PSKs: \n{}\n* messages:\n{}\n",
-            self.identifier(),
+            self.identifier().unwrap_or_default(),
```
Comment (arnauorriols) [inline streams/src/api/user.rs:1216]: not what I meant... and to be honest I liked better what you had originally 😅

File: streams/src/api/user.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnauorriols) [reply streams/src/api/user.rs]: Fair enough. I wonder though if having to call `to_identifier()` isn't as unclean as having to call `ok_or()`. But nevermind

**Review by arnauorriols — APPROVED** [review]
Ok, for me is a go. A couple of comments still open in case you guys want to follow up on them, but they can be dismissed all the same.

**Review by kwek20 — APPROVED** [review]

**Pull request merged**